### PR TITLE
add EC2 resource filter for disableApiTermination attribute

### PR DESCRIPTION
**Pull request by jantman**
This adds a filter to the ``ec2`` resource type for the ``disableApiTermination`` (a.k.a. Termination Protection) instance attribute.

The use case of ours for this is trying to terminate instances that have been stopped for greater than 60 days, but not forcing termination in case (as awful as this is) termination protection was enabled for a legitimate reason. So we end up with two policies like (note I've left out a lot of the notify fields for ease of reading):

```yaml
name: ec2-old-stopped-terminate
comment: EC2 Instances stopped longer than 60 days
resource: ec2
filters:
  - {State.Name: stopped}
  - type: state-age
    op: ge
    days: 60
  - not:
    - disableApiTermination
actions:
  - terminate
  - type: notify
    subject: '[cloud-custodian {{ account }}] EC2 Instances Terminated After Being Stopped for 60+ Days in {{ region }}'
```

and for those with termination protection:

```yaml
name: ec2-old-stopped-termination-protected
comment: EC2 Instances stopped longer than 60 days with termination protection enabled
resource: ec2
filters:
  - {State.Name: stopped}
  - type: state-age
    op: ge
    days: 60
  - disableApiTermination
actions:
  - type: notify
    subject: '[cloud-custodian {{ account }}] ACTION REQUIRED: EC2 Instances Stopped for 60+ Days WITH TERMINATION PROTECTION in {{ region }}'
```

### CODE REVIEW

**Comment by kapilt** [comment]
re filter name how about `termination-protected` ? we've generally gone lower case with hyphen for filter names. also, note this does require an api call per instance. also please sanitize out any account ids in recorded outputs.

**Comment by jantman** [comment]
@kapilt ok, will do about the name.

Understood on the API call per instance. I'll be honest I'm not sure about the internals of how filters like these are applied, but I was hoping c7n uses short-circuit evaluation (i.e. this will only be checked if earlier filters in the list pass?)

I'll remove the account IDs, but I really wasn't aware that just the IDs were a major concern...

**Comment by jantman** [comment]
@kapilt fixes pushed.

**Comment by kapilt** [comment]
account ids are vectors for searching for resources that are misconfigured, we've tried to adopt a general rule of only having known test accounts in recorded data.  If we look at the constant stream of s3 misconfigured data leaks in the news, thats in part because s3 is a global namespace without account id so you just need the resource name to probe. With an account id, you can dictionary/ngram probe for other resources that have embeddd iam policies that are misconfigured (lambda, kms, ecr, etc), since they require the resource name and account id for the probe.

**Comment by jantman** [comment]
Ahhh ok, makes sense. I'd used an empty lab account for these tests... but yeah, that sounds like good practice.

File: c7n/resources/ec2.py [hunk 1]
```diff
@@ -77,7 +77,7 @@ class resource_type(object):
     action_registry = actions
 
     # if we have to do a fallback scenario where tags don't come in describe
-    permissions = ('ec2:DescribeTags',)
+    permissions = ('ec2:DescribeTags', 'ec2:DescribeInstanceAttribute',)
```
Comment (kapilt) [inline c7n/resources/ec2.py]: hmm actually this permission should be on the disable filter not on the resource manager.

File: c7n/resources/ec2.py [hunk 2]
```diff
@@ -302,6 +302,54 @@ def __call__(self, i):
...
+            resource: ec2
+            filters:
+              - not:
+                - type: termination-protected
+    """
+
+    schema = type_schema('termination-protected')
+
+    def get_permissions(self):
+        return self.manager.get_resource_manager('ec2').get_permissions()
```
Comment (kapilt) [inline c7n/resources/ec2.py]: given above comment... re perms on this filter. then here. you can do perms = [self.permissions] and then perms.extend(self.manager.get_resource_manager('ec2').get

File: c7n/resources/ec2.py [hunk 3]
```diff
@@ -302,6 +302,54 @@ def __call__(self, i):
...
+
+    def is_termination_protection_enabled(self, inst):
+        manager = self.manager.get_resource_manager('ec2')
```
Comment (kapilt) [inline c7n/resources/ec2.py]: self.manager == ec2 manager.. its only when your cross across resource types that you really need to do the get_resource_manager call

File: c7n/resources/ec2.py (hunk 1 above)
Comment (jantman) [reply c7n/resources/ec2.py]: Oops, sorry about that. I was originally taking the approach of adding this code to DescribeEC2.augment() until I realized that DescribeInstanceAttribute needs to be called once per instance, and that would probably be a _really_ bad idea to do for every instance always.

File: c7n/resources/ec2.py (hunk 3 above)
Comment (jantman) [reply c7n/resources/ec2.py]: Ah ok, thanks.

**Comment by jantman** [comment]
@kapilt Sorry for what should've been a small change taking up so much of your time. I've made (I think) all of the changes that you requested, and also rebased on current master.

File: c7n/resources/ec2.py [hunk 4]
```diff
@@ -302,6 +302,56 @@ def __call__(self, i):
...
+        policies:
+          - name: termination-protection-NOT-enabled
+            resource: ec2
+            filters:
+              - not:
+                - type: termination-protected
+    """
+
+    schema = type_schema('termination-protected')
+    permissions = ['ec2:DescribeInstanceAttribute']
```
Comment (kapilt) [inline c7n/resources/ec2.py]: so this should to be a tuple, else when your extending below this list gets larger on every use of the filter within a given cli execution, (aka its like a mutable default variable that gets bigger on every use). and then below you can create a new list from the tuple to ensure its just the values needed.

File: c7n/resources/ec2.py [hunk 5]
```diff
@@ -302,6 +302,56 @@ def __call__(self, i):
...
+
+    def get_permissions(self):
+        perms = self.permissions
```
Comment (kapilt) [inline c7n/resources/ec2.py]: so to keep from extending the mutable class variable this needs to be turned into a new list reference, via 
`perms = list(self.permissions)`

File: c7n/resources/ec2.py (hunk 4 above)
Comment (jantman) [reply c7n/resources/ec2.py]: Oops, sorry.

**Comment by jantman** [comment]
@kapilt Fixes made per your last comments, and updated with latest master.

**Comment by kapilt** [comment]
thanks for your perseverance, lgtm

**Pull request merged**